### PR TITLE
Update Jupyter UI to latest version

### LINKED ISSUE #10: kubeflow 1.3 not creating gpu notebooks
I was able to get a kubeflow 1.3 installation using microk8s 1.21 and the steps documented here (with a --purge of microk8s/juju before):  https://github.com/ubuntu/microk8s/issues/2253#issuecomment-838158746

However, when creating a new notebook, I get the following error:

Screen Shot 2021-05-11 at 11 40 03 AM

Vanilla config, eg:

Screen Shot 2021-05-11 at 11 40 15 AM

### PR DESCRIPTION
**Pull request by knkski**
Updates Jupyter UI to latest version with UI revamp based on crud-web-apps library:

https://github.com/kubeflow/kubeflow/tree/master/components/crud-web-apps/jupyter

Also includes updated and enhanced tests that require deploying more charm dependencies, but allow actually testing that a notebook properly spins up.

Fixes #10

### CODE REVIEW

File: charms/jupyter-ui/metadata.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@ resources:
     # TODO(knkski): Switch back to upstream after this is merged:
     # https://github.com/kubeflow/kubeflow/pull/5686
```
Comment (DomFleischmann) [inline charms/jupyter-ui/metadata.yaml:14]: With this being the new UI mentioned in the issue do we still need our own images?

File: tests/test_selenium.py [hunk 2]
```diff
@@ -38,37 +40,73 @@ def driver(request):
...
-    # Ensure that the Notebook resource is created. It won't finish setting up
-    # due to a missing `default-editor` ServiceAccount that it relies on that's
-    # created by the kubeflow-profiles service.
-    row = wait.until(
-        lambda d: next(
-            row for row in d.find_elements_by_css_selector('.mat-row') if notebook_name in row.text
+    # Click submit on the form. Sleep for 1 second before clicking the submit
+    # button because shiny animations that ignore click events are simply a must.
+    # Note that that was sarcasm. If you're reading this, please don't shit up
+    # the web with braindead technologies.
```
Comment (DomFleischmann) [inline tests/test_selenium.py:61]: Noted.

File: tests/test_selenium.py [hunk 3]
```diff
@@ -38,37 +40,73 @@ def driver(request):
...
-            if f'check_circle\n{notebook_name}' in row.text
-        )
+    sleep(1)
+    submit.click()
+    wait.until(EC.url_to_be(url))
+
+    # Since upstream doesn't use proper class names or IDs or anything, find the
+    # <tr> containing elements that contain the notebook name and `ready`, signifying
+    # that the notebook is finished booting up. Returns a reference to the containing
+    # <tr> element. The result is a fairly unreadable XPath reference, but it works 🤷
```
Comment (DomFleischmann) [inline tests/test_selenium.py:74]: Not sure if we want to start having emojis in our codebase

File: .github/workflows/integrate.yaml [hunk 4]
```diff
@@ -50,11 +53,23 @@ jobs:
...
+        kubectl patch role -n kubeflow istio-ingressgateway-operator -p '{"apiVersion":"rbac.authorization.k8s.io/v1","kind":"Role","metadata":{"name":"istio-ingressgateway-operator"},"rules":[{"apiGroups":["*"],"resources":["*"],"verbs":["*"]}]}'
+        juju deploy cs:kubeflow-dashboard --config profile=kubeflow-user
+        juju deploy cs:admission-webhook
+        juju deploy cs:kubeflow-profiles
+        juju relate kubeflow-dashboard kubeflow-profiles
+
+        # https://bugs.launchpad.net/juju/+bug/1921739
+        (i=60; while ! juju wait -wvt 300 ; do ((--i)) || exit; sleep 1; done)
+
+        juju model-config update-status-hook-interval=5m
```
Comment (DomFleischmann) [inline .github/workflows/integrate.yaml:72]: This should probably be converted to pytest-operator.

**Review by DomFleischmann — CHANGES_REQUESTED** [review]

File: charms/jupyter-ui/metadata.yaml (hunk 1 above)
Comment (knkski) [reply charms/jupyter-ui/metadata.yaml:14]: Good catch, removed

File: .github/workflows/integrate.yaml (hunk 4 above)
Comment (knkski) [reply .github/workflows/integrate.yaml:72]: Mind if I kick that off to another PR? This is already a little bigger than I was planning

File: tests/test_selenium.py (hunk 3 above)
Comment (knkski) [reply tests/test_selenium.py:74]: We already do here: https://github.com/canonical/bundle-kubeflow/blob/master/tests/test_selenium.py#L22

Incidentally also added to placate my annoyance with having to deal with broken web technologies

File: .github/workflows/integrate.yaml (hunk 4 above)
Comment (DomFleischmann) [reply .github/workflows/integrate.yaml:72]: Sure!

File: charms/jupyter-ui/metadata.yaml (hunk 1 above)
Comment (DomFleischmann) [reply charms/jupyter-ui/metadata.yaml]: You removed the comment but left the rocks image, should we also switch to the upstream image?

Comment (knkski) [reply charms/jupyter-ui/metadata.yaml]: The upstream image in `gcr.io/kubeflow-images-public/jupyter-web-app` is out-of-date compared to what's in git master. That image registry is also no longer being updated. The new registry (`public.ecr.aws/j1r0q0g6/notebooks/jupyter-web-app`) [doesn't allow listing tags](https://github.com/kubeflow/kubeflow/issues/5922), so there's no way to see if that has an updated version. The upstream Kubeflow manifests repo only [shows](https://github.com/kubeflow/manifests/blob/master/apps/jupyter/jupyter-web-app/upstream/base/kustomization.yaml#L26) `v1.3.0rc1` from that registry.

Given all that, I just pulled down the latest version from git, built it, and pushed it up with the git hash as the tag. I think that's the best course of action unless you can see a good way around any of the above.

Comment (DomFleischmann) [reply charms/jupyter-ui/metadata.yaml]: Sounds good to me!

**Review by DomFleischmann — APPROVED** [review]

**Review by DomFleischmann — APPROVED** [review]

**Pull request merged**